google/cloud/storage/tests/object_media_integration_test.cc

```diff
@@ -56,6 +56,149 @@ bool UsingTestbench() {
+  std::string actual(std::istreambuf_iterator<char>{stream}, {});
+  ASSERT_FALSE(actual.empty());
+  auto expected_str = expected.str();
+  ASSERT_EQ(expected_str.size(), actual.size()) << " meta=" << meta;
+  EXPECT_EQ(expected_str, actual);
+
+  client.DeleteObject(bucket_name, object_name);
+  std::remove(file_name.c_str());
+}
+
+TEST_F(ObjectMediaIntegrationTest, DownloadFileFailure) {
+  Client client;
+  auto bucket_name = ObjectMediaTestEnvironment::bucket_name();
+  auto object_name = MakeRandomObjectName();
+  auto file_name = MakeRandomObjectName();
+
+#if GOOGLE_CLOUD_CPP_HAVE_EXCEPTIONS
+  EXPECT_THROW(
+      try {
+        client.DownloadToFile(bucket_name, object_name, file_name);
+      } catch (std::runtime_error const& ex) {
+        EXPECT_THAT(ex.what(), HasSubstr(object_name));
+        throw;
+      },
+      std::runtime_error);
+#else
+  EXPECT_DEATH_IF_SUPPORTED(
+      client.DownloadToFile(bucket_name, object_name, file_name),
+      "exceptions are disabled");
+#endif  // GOOGLE_CLOUD_CPP_HAVE_EXCEPTIONS
+}
+
+TEST_F(ObjectMediaIntegrationTest, DownloadFileCannotOpenFile) {
+  Client client;
+  auto bucket_name = ObjectMediaTestEnvironment::bucket_name();
+  auto object_name = MakeRandomObjectName();
+  ObjectMetadata meta =
+      client.InsertObject(bucket_name, object_name, LoremIpsum(),
+                          IfGenerationMatch(0), Projection::Full());
+
+  // Create an invalid path for the destination object.
+  auto file_name = MakeRandomObjectName() + "/" + MakeRandomObjectName();
+
+#if GOOGLE_CLOUD_CPP_HAVE_EXCEPTIONS
+  EXPECT_THROW(
+      try {
+        client.DownloadToFile(bucket_name, object_name, file_name);
+      } catch (std::runtime_error const& ex) {
+        EXPECT_THAT(ex.what(), HasSubstr(file_name));
+        throw;
+      },
+      std::runtime_error);
+#else
+  EXPECT_DEATH_IF_SUPPORTED(
+      client.DownloadToFile(bucket_name, object_name, file_name),
+      "exceptions are disabled");
+#endif  // GOOGLE_CLOUD_CPP_HAVE_EXCEPTIONS
+  client.DeleteObject(bucket_name, object_name);
+}
+
+TEST_F(ObjectMediaIntegrationTest, DownloadFileCannotWriteToFile) {
+#if GTEST_OS_LINUX
+  Client client;
+  auto bucket_name = ObjectMediaTestEnvironment::bucket_name();
+  auto object_name = MakeRandomObjectName();
+  ObjectMetadata meta =
+      client.InsertObject(bucket_name, object_name, LoremIpsum(),
+                          IfGenerationMatch(0), Projection::Full());
+
+  // We want to test that the code handles write errors *after* the file is
+  // successfully opened for writing. Such errors are hard to get, typically
+  // they indicate that the filesystem is full (or maybe some rare condition
+  // with remote filesystem such as NFS).
+  // On Linux we are fortunate that `/dev/full` meets those requirements
+  // exactly:
+  //   http://man7.org/linux/man-pages/man4/full.4.html
+  // I (coryan@) did not know about it, so I thought a longer comment may be in
```

Comment:
Super cool! Thanks for adding the comment.

Reply:
Don't feel like you need to justify long comments, especially for obscure stuff like this :P I almost always prefer more verbosity, since it usually reduces [WTFs/min](http://www.osnews.com/story/19266/WTFs_m)